pkg/utils/environment/kernel_symbols.go

```diff
@@ -2,319 +2,284 @@ package environment
-	symByAddr     map[addrAndOwner][]*KernelSymbol
-	requiredSyms  map[string]struct{}
-	requiredAddrs map[uint64]struct{}
-	onlyRequired  bool
-	updateLock    sync.Mutex
-	updateWg      sync.WaitGroup
-}
 
-func symNotFoundErr(v interface{}) error {
-	return fmt.Errorf("symbol not found: %v", v)
+// kernelSymbolInternal is a memory efficient representation of
+// a kernel symbol, used internally for storing all symbols.
+type kernelSymbolInternal struct {
+	name string
+	// We save only the low 48 bits of the address, as all (non-percpu) symbols are at 0xffffXXXXXXXXXXXX
+	// Owner is a 16-bit index into a slice of seen owners for the symbol table this symbol belongs to.
+	// It can only be translated to the owner name if we have the symbol table.
+	// To conserve memory, we encode both of them as a single 64-bit integer where the lower 48-bits
+	// are the address and the hight 16-bits are the owner index.
+	addressAndOwner uint64
 }
 
-// NewKernelSymbolTable initializes a KernelSymbolTable with optional configuration functions.
-func NewKernelSymbolTable(opts ...KSymbTableOption) (*KernelSymbolTable, error) {
-	k := &KernelSymbolTable{}
-	for _, opt := range opts {
-		if err := opt(k); err != nil {
-			return nil, err
-		}
-	}
-
-	// Set onlyRequired to true if there are required symbols or addresses
-	k.onlyRequired = k.requiredAddrs != nil || k.requiredSyms != nil
-
-	// Initialize maps if they are nil
-	if k.requiredSyms == nil {
-		k.requiredSyms = make(map[string]struct{})
-	}
-	if k.requiredAddrs == nil {
-		k.requiredAddrs = make(map[uint64]struct{})
+func newKernelSymbolInternal(name string, address uint64, owner uint16) *kernelSymbolInternal {
+	return &kernelSymbolInternal{
+		name:            name,
+		addressAndOwner: (uint64(owner) << 48) | (address & ((1 << 48) - 1)),
 	}
+}
 
-	return k, k.Refresh()
+func (ks kernelSymbolInternal) Name() string {
+	return ks.name
 }
 
-// KSymbTableOption defines a function signature for configuration options.
-type KSymbTableOption func(k *KernelSymbolTable) error
+func (ks kernelSymbolInternal) Address() uint64 {
+	// Convert truncated address to the real kernel address
+	return (0xffff << 48) | (ks.addressAndOwner & ((1 << 48) - 1))
+}
 
-// WithRequiredSymbols sets the required symbols for the KernelSymbolTable.
-func WithRequiredSymbols(reqSyms []string) KSymbTableOption {
-	return func(k *KernelSymbolTable) error {
-		k.requiredSyms = sliceToValidationMap(reqSyms)
-		return nil
-	}
+func (ks kernelSymbolInternal) owner() uint16 {
+	return uint16(ks.addressAndOwner >> 48)
 }
 
-// WithRequiredAddresses sets the required addresses for the KernelSymbolTable.
-func WithRequiredAddresses(reqAddrs []uint64) KSymbTableOption {
-	return func(k *KernelSymbolTable) error {
-		k.requiredAddrs = sliceToValidationMap(reqAddrs)
-		return nil
-	}
+func (ks kernelSymbolInternal) Contains(address uint64) bool {
+	return ks.Address() <= address && ks.Address()+maxSymbolSize > address
```

Comment:
consider calling it once:

```suggestion
	addr := ks.Address()
	return addr <= address && addr+maxSymbolSize > address
```